src/server/project.ts

```diff
@@ -1194,6 +1151,27 @@ namespace ts.server {
         }
     }
 
+    function getUnresolvedImports(program: Program, cachedUnresolvedImportsPerFile: Map<ReadonlyArray<string>>): SortedReadonlyArray<string> {
+        const ambientModules = program.getTypeChecker().getAmbientModules().map(mod => stripQuotes(mod.getName()));
+        return toDeduplicatedSortedArray(flatMap(program.getSourceFiles(), sourceFile =>

```

Comment:
flatMap will never return undefined (since there will be source files) and thus we are allocating single length array in all these updates.

Reply:
As of #28092 it should return the constant `emptyArray` instead of a new one.